### PR TITLE
Fix KBucket.add for a duplicate node of replacement_cache.

### PR DESCRIPTION
**Pull request by zsaladin**
### What was wrong?
- `KBucket.add` does not handle a duplicate node of `KBucket.replacement_cache`.

### How was it fixed?
- Add codes similar to handling duplicate nodes in `KBucket.nodes`

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: tests/p2p/test_kademlia.py [hunk 1]
```diff
@@ -114,6 +114,18 @@ def test_kbucket_add():
...
+    node4 = random_node()
+    assert bucket.add(node4) == node2
+    assert bucket.nodes == [node2, node]
+    assert bucket.head == node2
+    assert bucket.replacement_cache == [node3, node4]
+
+    assert bucket.add(node3) == node2
+    assert bucket.nodes == [node2, node]
+    assert bucket.head == node2
+    assert bucket.replacement_cache == [node4, node3]
```
Comment (zsaladin) [inline tests/p2p/test_kademlia.py]: Original codes works like below.
bucket.replacement_cache == [node3, node4, node3]

**Review by pipermerriam — APPROVED** [review]
It'd be nice to have a specific test that calls `bucket.remove()` since it's the simplest example.  :+1: to the current test but it mostly tests this using side effects and having at least one simple direct test would be nice (ideally as a completely independent test)

File: tests/p2p/test_kademlia.py [hunk 2]
```diff
@@ -144,6 +146,49 @@ def test_kbucket_remove():
...
+    assert bucket.nodes == list(reversed(nodes))
+    assert bucket.replacement_cache == replacement_nodes[1:] + replacement_nodes[:1]
+
+    for replacement_node in reversed(replacement_nodes):
+        assert bucket.add(replacement_node) == nodes[-1]
+    assert bucket.nodes == list(reversed(nodes))
+    assert bucket.replacement_cache == list(reversed(replacement_nodes))
+
+    all_nodes = list(itertools.chain(nodes, replacement_nodes))
+    random.shuffle(all_nodes)
```
Comment (carver) [inline tests/p2p/test_kademlia.py]: Having the test depend on `random` in this way would make it hard to reproduce if there was a particular ordering that was problematic. `hypothesis` is a good tool for this kind of thing. I think these are the main properties we are looking for in these particular tests:
- after the bucket is full, a newly added node is always at the end of the `replacement_cache`
- at all times, the `replacement_cache` is unique, ie~ `len(set(bucket.replacement_cache)) == len(bucket.replacement_cache)`

If you just throw a list of random nodes as input via hypothesis and meet those criteria ^ I think we are showing that the method is working correctly. That should make for a shorter, easier-to-read test.

(BTW, I see you aren't the first to introduce `random` into these tests, but let's not continue the old mistake :))

**Review by carver — COMMENTED** [review]
@zsaladin this change is small enough that it would be fine to add the `bucket.remove()` updates to the same PR. That will make for some more interesting hypothesis tests as well, like: add and then remove never leaves that node remaining in `.nodes` or `.replacement_cache`.

Looks like it needs a rebase, too.

**Comment by carver** [comment]
Oh and chances are good that the `cli` test failure was just our CI being wacky before and you shouldn't have problems on the next force-push.

**Pull request closed without merging**